gadgets/chaos_packet_drop/program.bpf.c

```diff
@@ -0,0 +1,385 @@
+			return TC_ACT_OK;
+
+		if (ingress && filter_ip.addr_raw.v4 != 0 &&
+		    (filter_ip.addr_raw.v4 != ip4h->saddr))
+			return TC_ACT_OK;
+		if (egress && filter_ip.addr_raw.v4 != 0 &&
+		    (filter_ip.addr_raw.v4 != ip4h->daddr))
+			return TC_ACT_OK;
+
+		event.src.addr_raw.v4 = ip4h->saddr;
+		key.dst.addr_raw.v4 = ip4h->daddr;
+		event.src.version = key.dst.version = 4;
+		sockets_key_for_md.family = SE_AF_INET;
+
+		switch (ip4h->protocol) {
+		case IPPROTO_TCP:
+			if (!filter_tcp)
+				return TC_ACT_OK;
+			struct tcphdr *tcph =
+				(struct tcphdr *)((__u8 *)ip4h +
+						  (ip4h->ihl * 4));
+			if ((void *)(tcph + 1) > data_end)
+				return TC_ACT_OK;
+
+			if (ingress && port != 0 &&
+			    port != bpf_ntohs(tcph->source))
+				return TC_ACT_OK;
+			if (egress && port != 0 &&
+			    port != bpf_ntohs(tcph->dest))
+				return TC_ACT_OK;
+
+			event.src.proto_raw = key.dst.proto_raw = IPPROTO_TCP;
+			event.src.port = bpf_ntohs(tcph->source);
+			key.dst.port = skb->pkt_type == SE_PACKET_HOST ?
+					       bpf_ntohs(tcph->dest) :
+					       bpf_ntohs(tcph->source);
+			sockets_key_for_md.proto = IPPROTO_TCP;
+			break;
+
+		case IPPROTO_UDP:
+			if (!filter_udp)
+				return TC_ACT_OK;
+			struct udphdr *udph =
+				(struct udphdr *)((__u8 *)ip4h +
+						  (ip4h->ihl * 4));
+			if ((void *)(udph + 1) > data_end)
+				return TC_ACT_OK;
+
+			if (ingress && port != 0 &&
+			    port != bpf_ntohs(udph->source))
+				return TC_ACT_OK;
+			if (egress && port != 0 &&
+			    port != bpf_ntohs(udph->dest))
+				return TC_ACT_OK;
+
+			event.src.port = bpf_ntohs(udph->source);
+			key.dst.port = skb->pkt_type == SE_PACKET_HOST ?
+					       bpf_ntohs(udph->dest) :
+					       bpf_ntohs(udph->source);
+			event.src.proto_raw = key.dst.proto_raw = IPPROTO_UDP;
+			sockets_key_for_md.proto = IPPROTO_UDP;
+			break;
+
+		default:
+			return TC_ACT_OK;
+		}
+		break;
+
+	case ETH_P_IPV6: // IPv6 Processing
+		if (filter_ip.version == 4 && filter_ip.addr_raw.v4 != 0)
+			return TC_ACT_OK;
+
+		ip6h = (struct ipv6hdr *)(eth + 1);
+
+		/* Check if IPv6 headers are invalid */
+		if ((void *)(ip6h + 1) > data_end)
+			return TC_ACT_OK;
+
+		if (ingress && filter_ip.version == 6 &&
+		    (__builtin_memcmp((const void *)filter_ip.addr_raw.v6,
+				      ip6h->saddr.in6_u.u6_addr8, 16)))
+			return TC_ACT_OK;
+		if (egress && filter_ip.version == 6 &&
+		    (__builtin_memcmp((const void *)filter_ip.addr_raw.v6,
+				      ip6h->daddr.in6_u.u6_addr8, 16)))
+			return TC_ACT_OK;
+
+		event.src.version = key.dst.version = 6;
+		sockets_key_for_md.family = SE_AF_INET6;
+		if (ip6h->nexthdr != NULL)
+			read_ipv6_address(&event, &key, ip6h);
+
+		switch (ip6h->nexthdr) {
+		case IPPROTO_TCP:
+			if (!filter_tcp)
+				return TC_ACT_OK;
+			struct tcphdr *tcph = (struct tcphdr *)(ip6h + 1);
+			if ((void *)(tcph + 1) > data_end)
+				return TC_ACT_OK;
+
+			if (ingress && port != 0 &&
+			    port != bpf_ntohs(tcph->source))
+				return TC_ACT_OK;
```

Comment:
nit: I'm not 100% sure, but perhaps it's better to apply the `bpf_htons` operation on the constant to avoid this bering performed for all packets. (Same in other places)